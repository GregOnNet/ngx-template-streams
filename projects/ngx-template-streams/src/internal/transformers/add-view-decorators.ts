import { flatten } from 'fp-ts/lib/Array';
import { fromNullable, some } from 'fp-ts/lib/Option';
import * as ts from 'typescript';
import {
  findNodes,
  findProperty,
  findPropertyByDecorator,
  getAllViewChildOptions,
  getComponentClass,
  getDecoratorArguments,
  getNamedImports,
  hasImportSpecifier
} from '../../utils/ast-helpers';
import { ObservableDecorator } from '../../utils/models';
import { INTERNAL_PREFIX } from '../constants';
import { getPropertyDecoratorQuery } from './queries';

enum ViewDecorator {
  ViewChild = 'ViewChild',
  ViewChildren = 'ViewChildren'
}

const ANGULAR_CORE_IMPORT = '@angular/core';

type DecoratorDeps = { [key in ObservableDecorator]: ViewDecorator };

const decoratorDeps: Partial<DecoratorDeps> = {
  ObservableChild: ViewDecorator.ViewChild,
  ObservableChildren: ViewDecorator.ViewChildren
};

export function addViewDecoratorsTransformer(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    return some(sourceFile)
      .chain(addImports(context))
      .chain(addProperties(context))
      .getOrElse(sourceFile);
  };
}

function addImports(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    const importsToAdd: ViewDecorator[] = getImportDependencies(sourceFile);

    const result = getNamedImports(sourceFile, ANGULAR_CORE_IMPORT).map(
      updateImports(importsToAdd, sourceFile, context)
    );

    return result.isNone() ? some(sourceFile) : result;
  };
}

function addProperties(context: ts.TransformationContext) {
  return (sourceFile: ts.SourceFile) => {
    return getComponentClass(sourceFile).map(updateClassDeclaration(sourceFile, context));
  };
}

function updateClassDeclaration(sourceFile: ts.SourceFile, context: ts.TransformationContext) {
  return (nodes: Array<ts.Node>) => {
    const visitor: ts.Visitor = (node: ts.Node) => {
      if (nodes.includes(node)) {
        const classDeclaration = node as ts.ClassDeclaration;

        const nodesByDecorator = Object.keys(decoratorDeps).map((decorator: ObservableDecorator) => {
          return findPropertyByDecorator(classDeclaration, decorator).map(property => {
            const [selector, , options] = getDecoratorArguments(property);

            const dep = decoratorDeps[decorator];
            const args = [selector];

            if (dep === ViewDecorator.ViewChild) {
              args.push(ts.createObjectLiteral(getQueryOptions(options)));
            }

            const propertyDecorator = ts.createDecorator(ts.createCall(ts.createIdentifier(dep), undefined, args));

            return ts.createProperty(
              [propertyDecorator],
              undefined,
              ts.createIdentifier(INTERNAL_PREFIX + property.name.getText()),
              undefined,
              undefined,
              undefined
            );
          });
        });

        const propertiesToAdd = flatten(nodesByDecorator);

        if (!propertiesToAdd.length) {
          return classDeclaration;
        }

        return ts.updateClassDeclaration(
          classDeclaration,
          classDeclaration.decorators,
          classDeclaration.modifiers,
          classDeclaration.name,
          classDeclaration.typeParameters,
          classDeclaration.heritageClauses,
          ts.createNodeArray([...classDeclaration.members, ...propertiesToAdd])
        );
      }

      return ts.visitEachChild(node, visitor, context);
    };

    return ts.visitNode(sourceFile, visitor);
  };
}

function updateImports(importsToAdd: ViewDecorator[], sourceFile: ts.SourceFile, context: ts.TransformationContext) {
  return (nodes: Array<ts.Node>) => {
    const visitor: ts.Visitor = (node: ts.Node) => {
      if (nodes.includes(node)) {
        const namedImports = node as ts.NamedImports;

        importsToAdd = importsToAdd.filter(i => !hasImportSpecifier(sourceFile, i));

        if (!importsToAdd.length) {
          return namedImports;
        }

        const newImportSpecifier = createImportSpecifier(importsToAdd);

        return ts.updateNamedImports(
          namedImports,
          ts.createNodeArray([...newImportSpecifier, ...namedImports.elements])
        );
      }

      return ts.visitEachChild(node, visitor, context);
    };

    if (!nodes.length) {
      const newImports = createImportSpecifier(importsToAdd);
      const namedImports = ts.createNamedImports(ts.createNodeArray([...newImports]));
      const importClause = ts.createImportClause(undefined, namedImports);

      const importDeclaration = ts.createImportDeclaration(
        undefined,
        undefined,
        importClause,
        ts.createLiteral(ANGULAR_CORE_IMPORT)
      );

      return ts.updateSourceFileNode(sourceFile, ts.createNodeArray([importDeclaration, ...sourceFile.statements]));
    }

    return ts.visitNode(sourceFile, visitor);
  };
}

function getImportDependencies(sourceFile: ts.SourceFile): ViewDecorator[] {
  const findDecorator = (decorator: ViewDecorator) => {
    return findNodes<ts.PropertyDeclaration>(sourceFile, getPropertyDecoratorQuery(decorator))
      .map(_ => decoratorDeps[decorator])
      .getOrElse(undefined);
  };

  return Object.keys(decoratorDeps)
    .map(findDecorator)
    .filter(Boolean);
}

function createImportSpecifier(identifiers: string[]) {
  return identifiers.map(name => ts.createImportSpecifier(undefined, ts.createIdentifier(name)));
}

function getQueryOptions(options: ts.Expression) {
  return fromNullable(options)
    .chain(options => {
      const allOptions = getAllViewChildOptions(options);
      const missingProperty = findProperty(options, 'static').filter(nodes => !nodes.length);

      return missingProperty.isSome()
        ? allOptions.map(userOptions => {
            const defaultOptions = getDefaultQueryOptions();
            return [...defaultOptions, ...userOptions];
          })
        : allOptions;
    })
    .getOrElse(getDefaultQueryOptions());
}

function getDefaultQueryOptions() {
  return [ts.createPropertyAssignment('static', ts.createFalse())];
}
