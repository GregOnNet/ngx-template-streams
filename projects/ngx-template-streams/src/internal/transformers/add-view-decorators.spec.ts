import { createFileContent } from '../../testing/test-helpers';
import { transformContent } from '../../testing/ts-compiler';
import * as transformer from './add-view-decorators';

describe('addViewDecoratorsTransformer', () => {
  let transformerSpy: jest.SpyInstance;

  beforeEach(() => {
    transformerSpy = jest.spyOn(transformer, 'addViewDecoratorsTransformer');
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  // no import declaration
  // import decl, but no view child / children
  // with other imports
  // multiple classes
  // test various query options

  it('should not apply transformation if the file is empty', () => {
    const content = '';

    const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

    expect(output).toMatchSnapshot();
    expect(transformerSpy).toHaveBeenCalled();
  });

  describe('imports', () => {
    describe('should not add import declaration', () => {
      test('when no component is defined', () => {
        const content = createFileContent`
          class TestComponent {}
        `;

        const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      });

      test('when decorator is not used', () => {
        const content = createFileContent`
          import { Component } from '@angular/core';

          @Component()
          class TestComponent {}
        `;

        const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      });
    });

    describe('when import declaration is missing', () => {
      test.each(['ObservableChild', 'ObservableChildren'])(
        'should add a new import declaration at the top › %s',
        decorator => {
          const content = createFileContent`
            import { bar } from 'some/path';

            @Component()
            class TestComponent {
              @${decorator}('foo', 'click')
              foo$: Observable<any>;
            }
          `;

          const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

          expect(output).toMatchSnapshot();
          expect(transformerSpy).toHaveBeenCalled();
        }
      );
    });

    it('should add ViewChild and ViewChildren import specifier', () => {
      const content = createFileContent`
        @Component()
        class TestComponent {
          @ObservableChild('foo', 'click')
          foo$: Observable<any>;

          @ObservableChildren('foo', 'click')
          bar$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(output).toContain('ViewChild');
      expect(output).toContain('ViewChildren');
      expect(transformerSpy).toHaveBeenCalled();
    });

    describe('when core import exists', () => {
      test.each(['ObservableChild', 'ObservableChildren'])(
        'should add import specifier to existing import declaration › %s',
        decorator => {
          const content = createFileContent`
            import { Component } from '@angular/core';

            @Component()
            class TestComponent {
              @${decorator}('foo', 'click')
              foo$: Observable<any>;
            }
          `;

          const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

          expect(output).toMatchSnapshot();
          expect(output).toContain(decorator);
          expect(transformerSpy).toHaveBeenCalled();
        }
      );

      test.each(['ObservableChild', 'ObservableChildren'])(
        'should add import specifier to an empty import statement › %s',
        decorator => {
          const content = createFileContent`
            import {} from '@angular/core';

            @Component()
            class TestComponent {
              @${decorator}('foo', 'click')
              foo$: Observable<any>;
            }
          `;

          const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

          expect(output).toMatchSnapshot();
          expect(transformerSpy).toHaveBeenCalled();
        }
      );

      test.each(['ObservableChild', 'ObservableChildren'])('should not change the order › %s', decorator => {
        const content = createFileContent`
          import { foo } from 'some/path';
          import { Component } from '@angular/core';
          import { of } from 'rxjs';

          @Component()
          class TestComponent {
            @${decorator}('foo', 'click')
            foo$: Observable<any>;
          }
        `;

        const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      });
    });

    describe('multiple classes', () => {
      it('should add imports only once', () => {
        const content = createFileContent`
          import { Component } from '@angular/core';

          @Component()
          class TestComponent {
            @ObservableChild('foo', 'click')
            foo$: Observable<any>;

            @ObservableChildren('bar', 'click')
            bar$: Observable<any>;
          }

          @Component()
          class TestComponent {
            @ObservableChild('foo', 'click')
            foo$: Observable<any>;

            @ObservableChildren('bar', 'click')
            bar$: Observable<any>;
          }
        `;

        const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      });

      it('should add imports needed for each class', () => {
        const content = createFileContent`
          import { Component } from '@angular/core';

          @Component()
          class TestComponent {
            @ObservableChild('foo', 'click')
            foo$: Observable<any>;
          }

          @Component()
          class TestComponent {
            @ObservableChildren('bar', 'click')
            bar$: Observable<any>;
          }
        `;

        const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

        expect(output).toMatchSnapshot();
        expect(transformerSpy).toHaveBeenCalled();
      });
    });
  });

  describe('ObservableChild', () => {
    it('should add default query options if not specified otherwise', () => {
      const content = createFileContent`
        import { Component } from '@angular/core';

        @Component()
        class TestComponent {
          @ObservableChild('foo', 'click')
          foo$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    it('should not overwrite user options', () => {
      const content = createFileContent`
        import { Component } from '@angular/core';

        @Component()
        class TestComponent {
          @ObservableChild('foo', 'click', { static: true, read: ElementRef })
          foo$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    it('should only apply valid query options', () => {
      const content = createFileContent`
        import { Component } from '@angular/core';

        @Component()
        class TestComponent {
          @ObservableChild('foo', 'click', { static: true, read: ElementRef, once: false, passive: true })
          foo$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    it('should work with component as selector', () => {
      const content = createFileContent`
        import { Component } from '@angular/core';

        @Component()
        class TestComponent {
          @ObservableChild(FooComponent, 'someOutput')
          foo$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });
  });

  describe('ObservableChildren', () => {
    it('should add property decorated with ViewChildren', () => {
      const content = createFileContent`
        import { Component } from '@angular/core';

        @Component()
        class TestComponent {
          @ObservableChildren('foo', 'click')
          foo$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });

    it('should add as many properties as observable children exist', () => {
      const content = createFileContent`
        import { Component } from '@angular/core';

        @Component()
        class TestComponent {
          @ObservableChildren('foo', 'click')
          foo$: Observable<any>;

          @ObservableChildren('bar', 'click')
          bar$: Observable<any>;

          @ObservableChildren(FooBarComponent, 'myOutput')
          foobar$: Observable<any>;
        }
      `;

      const output = transformContent(content, [transformer.addViewDecoratorsTransformer]);

      expect(output).toMatchSnapshot();
      expect(transformerSpy).toHaveBeenCalled();
    });
  });
});
