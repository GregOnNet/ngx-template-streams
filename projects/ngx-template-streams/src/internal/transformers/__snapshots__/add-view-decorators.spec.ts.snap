// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addViewDecoratorsTransformer ObservableChild should add default query options if not specified otherwise 1`] = `
"import { ViewChild, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer ObservableChild should not overwrite user options 1`] = `
"import { ViewChild, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click', { static: true, read: ElementRef })
    foo$: Observable<any>;
    @ViewChild('foo', { static: true, read: ElementRef })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer ObservableChild should only apply valid query options 1`] = `
"import { ViewChild, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click', { static: true, read: ElementRef, once: false, passive: true })
    foo$: Observable<any>;
    @ViewChild('foo', { static: true, read: ElementRef })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer ObservableChild should work with component as selector 1`] = `
"import { ViewChild, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild(FooComponent, 'someOutput')
    foo$: Observable<any>;
    @ViewChild(FooComponent, { static: false })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer ObservableChildren should add as many properties as observable children exist 1`] = `
"import { ViewChildren, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChildren('foo', 'click')
    foo$: Observable<any>;
    @ObservableChildren('bar', 'click')
    bar$: Observable<any>;
    @ObservableChildren(FooBarComponent, 'myOutput')
    foobar$: Observable<any>;
    @ViewChildren('foo')
    __foo$;
    @ViewChildren('bar')
    __bar$;
    @ViewChildren(FooBarComponent)
    __foobar$;
}
"
`;

exports[`addViewDecoratorsTransformer ObservableChildren should add property decorated with ViewChildren 1`] = `
"import { ViewChildren, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChildren('foo', 'click')
    foo$: Observable<any>;
    @ViewChildren('foo')
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports multiple classes should add imports needed for each class 1`] = `
"import { ViewChild, ViewChildren, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
}
@Component()
class TestComponent {
    @ObservableChildren('bar', 'click')
    bar$: Observable<any>;
    @ViewChildren('bar')
    __bar$;
}
"
`;

exports[`addViewDecoratorsTransformer imports multiple classes should add imports only once 1`] = `
"import { ViewChild, ViewChildren, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ObservableChildren('bar', 'click')
    bar$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
    @ViewChildren('bar')
    __bar$;
}
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ObservableChildren('bar', 'click')
    bar$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
    @ViewChildren('bar')
    __bar$;
}
"
`;

exports[`addViewDecoratorsTransformer imports should add ViewChild and ViewChildren import specifier 1`] = `
"import { ViewChild, ViewChildren } from \\"@angular/core\\";
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ObservableChildren('foo', 'click')
    bar$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
    @ViewChildren('foo')
    __bar$;
}
"
`;

exports[`addViewDecoratorsTransformer imports should not add import declaration when decorator is not used 1`] = `
"import { Component } from '@angular/core';
@Component()
class TestComponent {
}
"
`;

exports[`addViewDecoratorsTransformer imports should not add import declaration when no component is defined 1`] = `
"class TestComponent {
}
"
`;

exports[`addViewDecoratorsTransformer imports when core import exists should add import specifier to an empty import statement › ObservableChild 1`] = `
"import { ViewChild } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when core import exists should add import specifier to an empty import statement › ObservableChildren 1`] = `
"import { ViewChildren } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChildren('foo', 'click')
    foo$: Observable<any>;
    @ViewChildren('foo')
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when core import exists should add import specifier to existing import declaration › ObservableChild 1`] = `
"import { ViewChild, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when core import exists should add import specifier to existing import declaration › ObservableChildren 1`] = `
"import { ViewChildren, Component } from '@angular/core';
@Component()
class TestComponent {
    @ObservableChildren('foo', 'click')
    foo$: Observable<any>;
    @ViewChildren('foo')
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when core import exists should not change the order › ObservableChild 1`] = `
"import { foo } from 'some/path';
import { ViewChild, Component } from '@angular/core';
import { of } from 'rxjs';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when core import exists should not change the order › ObservableChildren 1`] = `
"import { foo } from 'some/path';
import { ViewChildren, Component } from '@angular/core';
import { of } from 'rxjs';
@Component()
class TestComponent {
    @ObservableChildren('foo', 'click')
    foo$: Observable<any>;
    @ViewChildren('foo')
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when import declaration is missing should add a new import declaration at the top › ObservableChild 1`] = `
"import { ViewChild } from \\"@angular/core\\";
import { bar } from 'some/path';
@Component()
class TestComponent {
    @ObservableChild('foo', 'click')
    foo$: Observable<any>;
    @ViewChild('foo', { static: false })
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer imports when import declaration is missing should add a new import declaration at the top › ObservableChildren 1`] = `
"import { ViewChildren } from \\"@angular/core\\";
import { bar } from 'some/path';
@Component()
class TestComponent {
    @ObservableChildren('foo', 'click')
    foo$: Observable<any>;
    @ViewChildren('foo')
    __foo$;
}
"
`;

exports[`addViewDecoratorsTransformer should not apply transformation if the file is empty 1`] = `""`;
